File: flutter_inappwebview_android/android/src/main/kotlin/dev/nosferatu500/inappwebview/webview/FlutterWebViewFactory.kt
package dev.nosferatu500.inappwebview.webview

import android.content.Context
import android.view.ViewGroup
import dev.nosferatu500.inappwebview.InAppWebViewFlutterPlugin
import dev.nosferatu500.inappwebview.webview.in_app_webview.FlutterWebView
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

// The unchecked cast below is the Flutter codec boundary: StandardMessageCodec decodes to
// Map<String,Object>/List<Object>, so every read of a structured value is an unverifiable
// cast. A wrong shape throws ClassCastException at the cast site, which is the intended
// failure mode.
@Suppress("UNCHECKED_CAST")
class FlutterWebViewFactory(private val plugin: InAppWebViewFlutterPlugin) :
  PlatformViewFactory(StandardMessageCodec.INSTANCE) {

  override fun create(context: Context, id: Int, args: Any?): PlatformView {
    val params = args as HashMap<String, Any?>
    var flutterWebView: FlutterWebView? = null
    var viewId: Any = id

    val keepAliveId = params["keepAliveId"] as String?
    val headlessWebViewId = params["headlessWebViewId"] as String?

    val headlessInAppWebViewManager = plugin.headlessInAppWebViewManager
    if (headlessWebViewId != null && headlessInAppWebViewManager != null) {
      val headlessInAppWebView = headlessInAppWebViewManager.webViews[headlessWebViewId]
      if (headlessInAppWebView != null) {
        flutterWebView = headlessInAppWebView.disposeAndGetFlutterWebView()
        flutterWebView?.keepAliveId = keepAliveId
      }
    }

    val inAppWebViewManager = plugin.inAppWebViewManager
    if (keepAliveId != null && flutterWebView == null && inAppWebViewManager != null) {
      flutterWebView = inAppWebViewManager.keepAliveWebViews[keepAliveId]
      if (flutterWebView != null) {
        // be sure to remove the view from the previous parent.
        val view = flutterWebView.getView()
        val parent = view?.parent as ViewGroup?
        parent?.removeView(view)
      }
    }

    val shouldMakeInitialLoad = flutterWebView == null
    if (flutterWebView == null) {
      if (keepAliveId != null) {
        viewId = keepAliveId
      }
      flutterWebView = FlutterWebView(plugin, context, viewId, params)
    }

    if (keepAliveId != null && inAppWebViewManager != null) {
      inAppWebViewManager.keepAliveWebViews[keepAliveId] = flutterWebView
    }

    if (shouldMakeInitialLoad) {
      flutterWebView.makeInitialLoad(params)
    }

    return flutterWebView
  }

  companion object {
    const val VIEW_TYPE_ID = "dev.nosferatu500.inappwebview/inappwebview"
  }
}

File: flutter_inappwebview_ios/ios/flutter_inappwebview_ios/Sources/flutter_inappwebview_ios/InAppWebView/FlutterWebViewFactory.swift
//
//  FlutterWebViewFactory.swift
//  flutter_inappwebview
//
//  Created by Lorenzo on 13/11/18.
//

import Flutter
import Foundation

/// `@MainActor` with an **isolated conformance** to Flutter's `FlutterPlatformViewFactory`
/// (SE-0470): `create(withFrame:viewIdentifier:arguments:)` builds a `FlutterWebViewController`,
/// touches `InAppWebViewManager.keepAliveWebViews` and calls `webView()` — all main-actor. Flutter
/// creates platform views on the platform thread, so this was already true.
@MainActor
public class FlutterWebViewFactory: NSObject, @MainActor FlutterPlatformViewFactory {
    static let VIEW_TYPE_ID = "dev.nosferatu500.inappwebview/inappwebview"
    
    private var plugin: InAppWebViewFlutterPlugin
    
    init(plugin: InAppWebViewFlutterPlugin) {
        self.plugin = plugin
        super.init()
    }
    
    public func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
    
    public func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        let arguments = args as? NSDictionary
        var flutterWebView: FlutterWebViewController?
        var id: Any = viewId
        
        let keepAliveId = arguments?["keepAliveId"] as? String
        let headlessWebViewId = arguments?["headlessWebViewId"] as? String
        let preventGestureDelay = arguments?["preventGestureDelay"] as? Bool ?? false
        
        if let headlessWebViewId = headlessWebViewId,
           let headlessWebView = plugin.headlessInAppWebViewManager?.webViews[headlessWebViewId],
           let platformView = headlessWebView?.disposeAndGetFlutterWebView(withFrame: frame) {
            flutterWebView = platformView
            flutterWebView?.keepAliveId = keepAliveId
        }
        
        if let keepAliveId = keepAliveId,
           flutterWebView == nil,
           let keepAliveWebView = plugin.inAppWebViewManager?.keepAliveWebViews[keepAliveId] {
            flutterWebView = keepAliveWebView
            if let view = flutterWebView?.view() {
                // remove from parent
                view.removeFromSuperview()
            }
        }
        
        let shouldMakeInitialLoad = flutterWebView == nil
        if flutterWebView == nil {
            if let keepAliveId = keepAliveId {
                id = keepAliveId
            }
            flutterWebView = FlutterWebViewController(plugin: plugin,
                                                      withFrame: frame,
                                                      viewIdentifier: id,
                                                      params: arguments!)
        }
        
        if let keepAliveId = keepAliveId {
            plugin.inAppWebViewManager?.keepAliveWebViews[keepAliveId] = flutterWebView!
        }
        
        flutterWebView?.webView()?.preventGestureDelay = preventGestureDelay
        
        if shouldMakeInitialLoad {
            flutterWebView?.makeInitialLoad(params: arguments!)
        }
        
        return flutterWebView!
    }
}
